File: sdks/ios/Sources/TaskirX/Network/RequestManager.swift
import Foundation

// MARK: - Custom Error

public enum TaskirXError: LocalizedError {
    case networkError(String)
    case decodingError(String)
    case httpError(statusCode: Int, message: String)
    case invalidResponse
    case timeout
    case retryExhausted
    
    public var errorDescription: String? {
        switch self {
        case .networkError(let message):
            return "Network Error: \(message)"
        case .decodingError(let message):
            return "Decoding Error: \(message)"
        case .httpError(let statusCode, let message):
            return "HTTP \(statusCode): \(message)"
        case .invalidResponse:
            return "Invalid response from server"
        case .timeout:
            return "Request timeout"
        case .retryExhausted:
            return "Max retries exhausted"
        }
    }
}

// MARK: - Request Manager

public class RequestManager {
    private let config: ClientConfig
    private let session: URLSession
    private var authToken: String?
    private let tokenLock = NSLock()
    
    public init(config: ClientConfig) {
        self.config = config
        
        let sessionConfig = URLSessionConfiguration.default
        sessionConfig.timeoutIntervalForRequest = config.timeout
        sessionConfig.timeoutIntervalForResource = config.timeout * 2
        self.session = URLSession(configuration: sessionConfig)
    }
    
    public func setAuthToken(_ token: String) {
        tokenLock.lock()
        defer { tokenLock.unlock() }
        self.authToken = token
    }
    
    public func clearAuthToken() {
        tokenLock.lock()
        defer { tokenLock.unlock() }
        self.authToken = nil
    }
    
    // MARK: - HTTP Methods
    
    public func get<T: Decodable>(_ endpoint: String) async throws -> T {
        var urlComponents = URLComponents(string: config.apiUrl + endpoint)!
        return try await request(urlComponents.url!, method: "GET")
    }
    
    public func post<T: Decodable>(_ endpoint: String, body: Encodable? = nil) async throws -> T {
        let url = URL(string: config.apiUrl + endpoint)!
        return try await request(url, method: "POST", body: body)
    }
    
    public func put<T: Decodable>(_ endpoint: String, body: Encodable? = nil) async throws -> T {
        let url = URL(string: config.apiUrl + endpoint)!
        return try await request(url, method: "PUT", body: body)
    }
    
    public func delete<T: Decodable>(_ endpoint: String) async throws -> T {
        let url = URL(string: config.apiUrl + endpoint)!
        return try await request(url, method: "DELETE")
    }
    
    // MARK: - Core Request Logic
    
    private func request<T: Decodable>(_ url: URL, method: String, body: Encodable? = nil) async throws -> T {
        var lastError: Error?
        let delays = [100, 300, 900] // milliseconds for exponential backoff
        
        for attempt in 0..<config.retryAttempts {
            do {
                var request = URLRequest(url: url)
                request.httpMethod = method
                request.timeoutInterval = config.timeout
                
                // Add headers
                request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                request.setValue("application/json", forHTTPHeaderField: "Accept")
                request.setValue(config.apiKey, forHTTPHeaderField: "X-API-Key")
                request.setValue(UUID().uuidString, forHTTPHeaderField: "X-Request-ID")
                request.setValue("TaskirX-iOS/1.0", forHTTPHeaderField: "User-Agent")
                
                // Add auth token if available
                if let token = authToken {
                    request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
                }
                
                // Add body if provided
                if let body = body {
                    request.httpBody = try JSONEncoder().encode(body)
                }
                
                let (data, response) = try await session.data(for: request)
                
                // Check HTTP status
                guard let httpResponse = response as? HTTPURLResponse else {
                    throw TaskirXError.invalidResponse
                }
                
                guard (200..<300).contains(httpResponse.statusCode) else {
                    if let errorResponse = try? JSONDecoder().decode(ErrorResponse.self, from: data) {
                        throw TaskirXError.httpError(statusCode: httpResponse.statusCode, message: errorResponse.message)
                    }
                    throw TaskirXError.httpError(statusCode: httpResponse.statusCode, message: "HTTP Error")
                }
                
                // Decode response
                let decoder = JSONDecoder()
                decoder.keyDecodingStrategy = .convertFromSnakeCase
                let decodedResponse = try decoder.decode(ApiResponse<T>.self, from: data)
                
                if decodedResponse.success, let data = decodedResponse.data {
                    if config.debug {
                        print("[TaskirX] ✅ \(method) \(url.path) - Success")
                    }
                    return data
                } else if let error = decodedResponse.error {
                    throw TaskirXError.httpError(statusCode: httpResponse.statusCode, message: error.message)
                } else {
                    throw TaskirXError.decodingError("Invalid response structure")
                }
                
            } catch let error as TaskirXError {
                lastError = error
                
                if attempt < config.retryAttempts - 1 {
                    let delayMs = delays[min(attempt, delays.count - 1)]
                    if config.debug {
                        print("[TaskirX] ⚠️ Retry attempt \(attempt + 1) after \(delayMs)ms")
                    }
                    try await Task.sleep(nanoseconds: UInt64(delayMs * 1_000_000))
                } else {
                    if config.debug {
                        print("[TaskirX] ❌ \(method) \(url.path) - \(error)")
                    }
                }
            } catch {
                lastError = TaskirXError.networkError(error.localizedDescription)
                
                if attempt < config.retryAttempts - 1 {
                    let delayMs = delays[min(attempt, delays.count - 1)]
                    if config.debug {
                        print("[TaskirX] ⚠️ Retry attempt \(attempt + 1) after \(delayMs)ms")
                    }
                    try await Task.sleep(nanoseconds: UInt64(delayMs * 1_000_000))
                }
            }
        }
        
        throw lastError ?? TaskirXError.retryExhausted
    }
}

File: sdks/react-native/src/network/RequestManager.ts
// HTTP Request Manager for React Native

let Platform: any;
try {
  Platform = require('react-native').Platform;
} catch {
  Platform = { OS: 'web', Version: '1.0' };
}

import type { ClientConfig, TaskirXError, ApiResponse } from './types';
import { TaskirXErrorType } from './types';

export class RequestManager {
  private config: ClientConfig;
  private authToken: string | null = null;
  private tokenLock = Promise.resolve();

  constructor(config: ClientConfig) {
    this.config = {
      timeout: 30000,
      retryAttempts: 3,
      ...config,
    };
  }

  setAuthToken(token: string): void {
    this.authToken = token;
  }

  clearAuthToken(): void {
    this.authToken = null;
  }

  private createError(
    type: TaskirXErrorType,
    message: string,
    statusCode?: number,
    originalError?: Error
  ): TaskirXError {
    return {
      type,
      message,
      statusCode,
      originalError,
    };
  }

  async get<T>(endpoint: string): Promise<T> {
    return this.request<T>(`${this.config.apiUrl}${endpoint}`, 'GET');
  }

  async post<T>(endpoint: string, body?: any): Promise<T> {
    return this.request<T>(`${this.config.apiUrl}${endpoint}`, 'POST', body);
  }

  async put<T>(endpoint: string, body?: any): Promise<T> {
    return this.request<T>(`${this.config.apiUrl}${endpoint}`, 'PUT', body);
  }

  async delete<T>(endpoint: string): Promise<T> {
    return this.request<T>(`${this.config.apiUrl}${endpoint}`, 'DELETE');
  }

  private async request<T>(
    url: string,
    method: string,
    body?: any
  ): Promise<T> {
    const delays = [100, 300, 900]; // milliseconds
    let lastError: TaskirXError | null = null;

    for (let attempt = 0; attempt < this.config.retryAttempts!; attempt++) {
      try {
        const headers = this.buildHeaders();
        const options: RequestInit & { timeout?: number } = {
        method,
        headers,
        timeout: this.config.timeout,
      };

        if (body && (method === 'POST' || method === 'PUT')) {
          options.body = JSON.stringify(body);
        }

        if (this.config.debug) {
          console.log(`[TaskirX] 🔵 ${method} ${url}`);
        }

        const response = await fetch(url, options);

        if (!response.ok) {
          const errorData = await response.json().catch(() => ({}));
          const errorMessage = errorData.message || `HTTP ${response.status}`;
          throw this.createError(
            TaskirXErrorType.HTTP_ERROR,
            errorMessage,
            response.status
          );
        }

        const data = await response.json();

        if (!data.success && data.error) {
          throw this.createError(
            TaskirXErrorType.HTTP_ERROR,
            data.error.message,
            response.status
          );
        }

        if (this.config.debug) {
          console.log(`[TaskirX] ✅ ${method} ${url} - Success`);
        }

        return data.data as T;
      } catch (error) {
        if (error instanceof Error && 'type' in error) {
          lastError = error as TaskirXError;
        } else if (error instanceof TypeError) {
          lastError = this.createError(
            TaskirXErrorType.NETWORK_ERROR,
            error.message,
            undefined,
            error
          );
        } else {
          lastError = this.createError(
            TaskirXErrorType.NETWORK_ERROR,
            'Unknown error occurred',
            undefined,
            error instanceof Error ? error : new Error(String(error))
          );
        }

        if (attempt < this.config.retryAttempts! - 1) {
          const delayMs = delays[Math.min(attempt, delays.length - 1)];
          if (this.config.debug) {
            console.log(`[TaskirX] ⚠️ Retry attempt ${attempt + 1} after ${delayMs}ms`);
          }
          await this.sleep(delayMs);
        } else {
          if (this.config.debug) {
            console.log(`[TaskirX] ❌ ${method} ${url} - ${lastError.message}`);
          }
        }
      }
    }

    throw lastError || this.createError(
      TaskirXErrorType.RETRY_EXHAUSTED,
      'Max retries exhausted'
    );
  }

  private buildHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      'X-API-Key': this.config.apiKey,
      'X-Request-ID': this.generateRequestId(),
      'User-Agent': this.buildUserAgent(),
    };

    if (this.authToken) {
      headers['Authorization'] = `Bearer ${this.authToken}`;
    }

    return headers;
  }

  private buildUserAgent(): string {
    const platform = Platform.OS === 'ios' ? 'iOS' : 
                    Platform.OS === 'android' ? 'Android' : 'Unknown';
    return `TaskirX-RN/1.0 (${platform}/${Platform.Version})`;
  }

  private generateRequestId(): string {
    return `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
  }

  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}
